### DISCUSSION TITLE
GDL as Flatpak

### DISCUSSION THREAD

**Discussion by amolh12** [gnudatalanguage/gdl#914] [General]
Dear All,

I just recently switched to OpenSuse from Fedora, and really struggling to get the GDL up and running on OpenSuse. It would be really great if GDL team please release it as a Flatpak / Snap application. The portability with all the Linux distro would be really great! 

Thanks

**Comment by slayoo** [comment]
Hello, 
Please to elaborate on what is causing problems on OpenSuse and how have you tried to get/compile it (version?).
@gnudatalanguage/gdlteam any one with experience in Flatpak / Snap packaging?
Thanks,
Sylwester

**Reply by amolh12** [reply]
Thanks for your reply. I used the git clone to download the gdl. After running **rpm4gdl.sh** I got, 

`Please wait ! need some time.
 
Missing Mandatory packages you need to install :
sudo rpm -i libpng-devel

Missing packages you may would like to install :
sudo rpm -i GraphicsMagick-c++-devel ImageMagick-c++-devel wxGTK-devel fftw-devel pslib-devel grib_api-devel
`
These packages are already installed but with different naming schemes, as per the OpenSuse repository, e.g. libpng16-devel is already installed and so on. 

After running the CMake in the build directory, errors occurred. Attaching the CMakeError.log and CMakeOutput.log files. 
[CMakeError.log](https://github.com/gnudatalanguage/gdl/files/6293737/CMakeError.log)
[CMakeOutput.log](https://github.com/gnudatalanguage/gdl/files/6293738/CMakeOutput.log)

As the GDL is actively maintained on Fedora, the creation of Flatpak from the Fedora Repository would be really straightforward task for expert developers. All the dependencies etc will be automatically bundled in a single Flatpak package. The following link might be really handy for the creation of the Flatpak from the Fedora repositories. 

https://docs.fedoraproject.org/en-US/flatpak/tutorial/

I am using OpenSuse Tumbleweed, output of neofetch is also attached. 
![os](https://user-images.githubusercontent.com/1918403/114338732-54b9bb00-9b71-11eb-9856-f82bf43cc409.png)

**Reply by slayoo** [reply]
Seems that plplot installation is not functional, the error log ends with:
```
*** PLPLOT ERROR, ABORTING OPERATION ***
plInitDispatchTable: Could not open drivers directory, aborting operation
```

**Reply by amolh12** [reply]
All the plplot libraries are installed, still the following error is happening..there are plenty of such errors about other packages as well even I compile with  -DINTERACTIVE_GRAPHICS=OFF as suggested. 

<pre>This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found PLPLOT: /usr/lib64/libplplot.so;/usr/lib64/libplplotcxx.so  
-- Warning, if you have plplot version &gt; 5.11 and wxWidgets enabled, please read carefully file README
-- Looking for c_plslabelfunc in /usr/lib64/libplplot.so;/usr/lib64/libplplotcxx.so
-- Looking for c_plslabelfunc in /usr/lib64/libplplot.so;/usr/lib64/libplplotcxx.so - found
-- Looking for c_plwidth in /usr/lib64/libplplot.so;/usr/lib64/libplplotcxx.so
-- Looking for c_plwidth in /usr/lib64/libplplot.so;/usr/lib64/libplplotcxx.so - found
-- Looking for plstrl in /usr/lib64/libplplot.so;/usr/lib64/libplplotcxx.so
-- Looking for plstrl in /usr/lib64/libplplot.so;/usr/lib64/libplplotcxx.so - found
-- Performing Test PLPLOT_HAS_PLCALLBACK
-- Performing Test PLPLOT_HAS_PLCALLBACK - Success
-- Performing Test HAVE_PLPLOT_WXWIDGETS_DRIVER
-- Performing Test HAVE_PLPLOT_WXWIDGETS_DRIVER - Failed
CMake Error at CMakeLists.txt:584 (message):
  PlPlot was compiled without needed wxwidgets driver, aborting.  Do you mean
  -DINTERACTIVE_GRAPHICS=OFF ?


-- Configuring incomplete, errors occurred!
</pre>

**Reply by GillesDuvert** [reply]
Hi @amolh12 , I 'm afraid you suffer from the result of our new building policy, although a flaw in our procedure is possible.

Basically, we absolutely need to know at build  if plplot is installed with the wxwidgets support, be it a dynamic or static driver. And **refuse to build if not**. This is imperative if we want the _packagers provide a working gdl and not something that will crash at the first plot_ because the plplot driver is not present.

The CMakeLists.txt provides cmake with a simple test to check that. Apparently it fails miserably:
```
*** PLPLOT ERROR, ABORTING OPERATION ***
plInitDispatchTable: Could not open drivers directory, aborting operation

Return value: 1
Source file was:

      #include <plplot/plplot.h>
      #include<string.h>
      int main(int argc, char **argv) {
      int numdevs = 128;
      const char **devlongnames = (const char**) malloc(numdevs * sizeof (char*));
      const char **devnames = (const char**) malloc(numdevs * sizeof (char*));
      plgDevs(&devlongnames, &devnames, &numdevs);
      for (int i = 0; i < numdevs; ++i) {if (strcmp("wxwidgets", devnames[i])==0) exit(EXIT_SUCCESS);}
      exit(EXIT_FAILURE);
      }
```
It was not intended to fail on `plInitDispatchTable: Could not open drivers directory, aborting operation` however.
I do not see why plplot would not "open drivers directory", is it because plplot is built without dynamic drivers? the plplot library is by construction very diverse, and has too many possibilities for us to handle. Could you check that plplot works (independently) and that it has installed e.g., the wxwidgets driver in something like `/usr/lib64/plplot5.14.0/drivers/wxwidgets.so` (5.14.0 in my mageia version). 
As the cmake is suposed to work, provided you have removed any Cmake Cache left, and type 
`cmake -DINTERACTIVE_GRAPHICS=OFF ... (etc)` it should not hang (if it does, then it's another bug). If it works however you'll not get any graphic outputs, that's probably not what you expect.

**Reply by amolh12** [reply]
Yes it is installed in 
`/usr/lib64/plplot5.15.0/drivers/wxwidgets.so`

**Reply by GillesDuvert** [reply]
Is it because you need to define PLPLOT_DRV_DIR before compiling something using plplot, [as suggested here](https://www.mail-archive.com/plplot-general@lists.sourceforge.net/msg01381.html).

**Comment by alaingdl** [comment]
Yes, it would be great to know the issues. I am ready to add somethings in the scripts/ dir. to help, I just received feedback for the Debian case.

Despite incredible administrative issues, I hope to have 2 interns this summer, and that is a very good topic :)

**Comment by PattiMichelle** [comment]
This may be stupid - but why not build statically-linked binary for "all
other distros?"   A robot-build once per month?

Example:
https://salome-platform.org/downloads/current-version  
(/see "universal linux binary" download/)

It seems like that would avoid complicated solutions.

**Reply by GillesDuvert** [reply]
... hopefully not 1.8 GBytes download :smile:

**Reply by GillesDuvert** [reply]
more seriously, provided "universal linux" really exist, it would be great indeed.

**Reply by amolh12** [reply]
Well, Flatpak is quite reasonable to bridge the gap among all the Linux distro. So GDL just need to be released as a Flatpak application, and can easily be hosted on 

https://flathub.org/home

**Comment by dirteat** [comment]
How about reporting a bug report to OpenSuse, that's just a distro bug, not a bug in GDL. This issue should be closed as "invalid".

About statistically linked binary, welcome back to the 80s? Why do you think shared library has been invented on operating system?

Releasing only a flatpak means that the source code and compilation system will be poorly tested, open door to bugs. It will also let grow the gap between API used in GDL and API provided by the most recent libraries GDL needs. At some point, to keep a binary version running, developpers will package obsolete frozen versions of these libraries within the binary, and this will be another open door to many bugs. At some point the package will be completely separated to all the mainstream libs it needs, and ready to die with a forgotten  code.

Check out how many code from the 80s are still running today... One exception being "iraf", but that's because it moved back to community and can start again be compiled against modern libs.

PS: that does not forbid to have a flatpak release, but that is more work.

Cheers.

**Comment by PattiMichelle** [comment]
I'm more than familiar with the philosophies associated - but this isn't
an OS.  That's why I included a link to a /fully-maintained /modern,
complex code which /does/ do a static binary in addition to distros -
there has been no devastating stuff associated with this practice.  It
appears to involve less work. 

Enjoy.

**Comment by amolh12** [comment]
I finally managed to run the GDL on the OpenSUSE. Following steps I took: 

Install GDL from Fedora repository.

`ldd /usr/bin/gdl | grep "=> /" | awk '{print $3}' | xargs -I '{}' cp -v '{}' ./gdl_lib/`

The above step will extract all the libraries required. 

On OpenSUSE: `$ zypper in plplot-driver-*`

Copy `/usr/bin/gdl` from Fedora along with the `gdl_lib`

`/lib64/ld-2.33.so --library-path /opt/GDL/gdl_lib  /opt/GDL/gdl-bin`

**Reply by amolh12** [reply]
any better/polished way to accomplish the same would be really appreciated.
